arangod/Aql/QueryCache.cpp

```diff
@@ -193,19 +331,71 @@ void QueryCacheDatabaseEntry::invalidate(std::string const& dataSource) {
 
 /// @brief enforce maximum number of results
 /// must be called under the shard's lock
-void QueryCacheDatabaseEntry::enforceMaxResults(size_t value) {
-  while (_numElements > value) {
+void QueryCacheDatabaseEntry::enforceMaxResults(size_t numResults, size_t sizeResults) {
+  while (_numResults > numResults || 
+         _sizeResults > sizeResults) {
     // too many elements. now wipe the first element from the list
 
     // copy old _head value as unlink() will change it...
     auto head = _head;
+    removeDatasources(head);
     unlink(head);
     auto it = _entriesByHash.find(head->_hash);
     TRI_ASSERT(it != _entriesByHash.end());
     _entriesByHash.erase(it);
   }
 }
 
+/// @brief enforce maximum per-entry size
+/// must be called under the shard's lock
+void QueryCacheDatabaseEntry::enforceMaxEntrySize(size_t value) {
+  for (auto it = _entriesByHash.begin(); it != _entriesByHash.end(); /* no hoisting */) {
+    auto const& entry = (*it).second.get();
+
+    if (entry->_size > value) {
+      removeDatasources(entry);
+      unlink(entry);
+      it = _entriesByHash.erase(it);
+    } else {
+      // keep the entry
+      ++it;
+    }
+  }
+}
+
+/// @brief exclude all data from system collections
+/// must be called under the shard's lock
+void QueryCacheDatabaseEntry::excludeSystem() {
+  for (auto it = _entriesByDataSource.begin(); it != _entriesByDataSource.end(); /* no hoisting */) {
+    if ((*it).first.empty() || (*it).first[0] != '_') {
+      // not a system collection
+      ++it;
+    } else {
+      for (auto const& hash : (*it).second) {
+        auto it2 = _entriesByHash.find(hash);
+        
+        if (it2 != _entriesByHash.end()) {
+          auto* entry = (*it2).second.get();
+          unlink(entry);
+          _entriesByHash.erase(it2);
+        }
+      }
+
+      it = _entriesByDataSource.erase(it);
+    }
+  }
+}
+
+void QueryCacheDatabaseEntry::removeDatasources(QueryCacheResultEntry const* e) {
+  for (auto const& ds : e->_dataSources) {
+    auto it = _entriesByDataSource.find(ds);
+
+    if (it != _entriesByDataSource.end()) {
```

Comment:
are empty map entries cleaned up somewhere ?

Reply:
They are cleaned in `QueryCacheDatabaseEntry::invalidate(std::string const& dataSource)`.